### ISSUE TITLE
Support environment variables as well as a config file for setup

### ISSUE THREAD

**Issue by davidgibbons** [vouch/vouch-proxy#132]
Starting to look into using vouch to secure some internal resources at work and it would be so much easier to integrate into my workflow if it could take it's configuration settings as  environment variables.  

We inject our secrets into the configuration via a third party tool as environment variables. In order to write out a valid config file i'm going to have to create a wrapper. Supporting things like client_id and client_secret as env vars would reduce the custom work I'd have to maintain.

It also supports more 12 factor style deployments.

**Comment by bnfinet** [comment]
Thanks for the suggestion @davidgibbons and welcome back :)

There's been talk on occasion about adding environmental variable based configuration with an eye on supporting a 12 factor setup.

The configuration library in use is [spf13/viper](https://github.com/spf13/viper#working-with-environment-variables) which does have support for such right out of the box...
https://github.com/spf13/viper#working-with-environment-variables

My hope is that it wouldn't be too hard to add.  If you're at all interested in hacking on this aspect of the code do let me know. 

Hope to see you around Portland.

**Comment by Coderrob** [comment]
@bnfinet there are some PR's in the Viper repo related to this. Ran into this issue while diving into whether a physical config file was a requirement to load just environment variables. It is, at least for the current 1.6.3v of Viper. 

I did come across an open PR in Viper https://github.com/spf13/viper/pull/723/files that addresses this by creating a physical file. Not ideal, when the Viper library ultimately is just loading a map of configs, and having a dependency on a physical file doesn't make a lot of sense for most applications. 

Anyway, just wanted to point that out.

**Comment by bnfinet** [comment]
thanks much @Coderrob 

I think I'm hearing that the `.defaults.yml` strategy outlined in #134 is still the best, most viable option (at least with using `spf13/viper`).  Do I have that right?

**Comment by bnfinet** [comment]
@davidgibbons or @Coderrob if you cared to kick the tires on this it would be much appreciated

Still have to update the README and `config.yml` so the best guide to env var naming is the `cfg_test.go` file or the struct in `pkg/cfg/cfg.go` and `pkg/cfg/oauth.go`

I'll add a few more tests for config overriding (ENV overwrites config file, etc) in the coming days but this is feeling pretty solid.

The Viper based solution was unfortunately not tenable.  I went a bit down the rabbit hole of something homegrown using reflection but after realizing I was reinventing [envconfig](https://github.com/kelseyhightower/envconfig) I went that way.
